### ISSUE TITLE
Error on accessing struct members after atomic load

### ISSUE THREAD

**Issue by wrenger** [llvm/llvm-project#66539]
The following code (compiled with clang version 16.0.6 (Fedora 16.0.6-3.fc38)) produces the error "cannot compile this l-value expression yet". GCC (13.2.1) does not complain and compiles this successfully.

I found other issues with this error message, but they did not seem to be related (from what I understand) and most focus on C++.

```sh
clang test.c -o test
gcc test.c -o test
```

```c
#include <stdatomic.h>
#include <stdio.h>

struct data {
  int inner;
};

int main() {
  _Atomic(struct data) data = (struct data){0};

  printf("%d", atomic_load(&data).inner);
}
```

Full error message:
```
test.c:11:16: error: cannot compile this l-value expression yet
  printf("%d", atomic_load(&data).inner);
               ^~~~~~~~~~~~~~~~~~
/usr/lib64/clang/16/include/stdatomic.h:133:29: note: expanded from macro 'atomic_load'
#define atomic_load(object) __c11_atomic_load(object, __ATOMIC_SEQ_CST)
                            ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
1 error generated.
```

**Comment by shafik** [comment]
Looks like this diagnostic is being emitted from here:

https://github.com/llvm/llvm-project/blob/401296f697b4ec2088afa0637d698999b2fe7228/clang/lib/CodeGen/CGExpr.cpp#L1287-L1291

So this looks like a codegen issue but I am not sure who the right person to pass this on to hopefully @AaronBallman @efriedma-quic might know.

**Comment by wrenger** [comment]
It seems that atomics are only implemented for assign expressions.

https://github.com/llvm/llvm-project/blob/401296f697b4ec2088afa0637d698999b2fe7228/clang/lib/CodeGen/CGExpr.cpp#L1302-L1305

**Comment by efriedma-quic** [comment]
Probably we need code to make codegen handle AtomicExpr the same way as a call that returns a struct, i.e. create a temporary.

You could maybe argue that Sema should construct a MaterializeTemporaryExpr like we do in C++ mode, but that seems like a big change for a minor benefit.

**Comment by efriedma-quic** [comment]
It's really just an edge case: the combination of atomics, and accessing members of rvalues. C doesn't doesn't really have temporaries like C++, but it still allows accessing a member of an rvalue.  So there are special cases to handle this in the few places where it comes up.  It looks like we just forgot to add such a special case for atomic builtins.
